File: src/main/java/mods/eln/PacketHandler.java
package mods.eln;

import cpw.mods.fml.common.eventhandler.SubscribeEvent;
import cpw.mods.fml.common.network.FMLNetworkEvent.ServerCustomPacketEvent;
import cpw.mods.fml.common.network.internal.FMLProxyPacket;
import io.netty.channel.ChannelHandler.Sharable;
import mods.eln.client.ClientProxy;
import mods.eln.debug.DP;
import mods.eln.debug.DPType;
import mods.eln.misc.Coordinate;
import mods.eln.misc.IConfigSharing;
import mods.eln.misc.KeyRegistry;
import mods.eln.misc.Utils;
import mods.eln.node.INodeEntity;
import mods.eln.node.NodeBase;
import mods.eln.node.NodeManager;
import mods.eln.server.PlayerManager;
import mods.eln.sound.ClientSoundHandler;
import mods.eln.sound.SoundClient;
import mods.eln.sound.SoundCommand;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.network.NetHandlerPlayServer;
import net.minecraft.network.NetworkManager;
import net.minecraft.tileentity.TileEntity;

import java.io.*;

@Sharable
public class PacketHandler {

    public PacketHandler() {
        Eln.eventChannel.register(this);
    }


    @SubscribeEvent
    public void onServerPacket(ServerCustomPacketEvent event) {
        FMLProxyPacket packet = event.packet;
        DataInputStream stream = new DataInputStream(new ByteArrayInputStream(packet.payload().array()));
        NetworkManager manager = event.manager;
        EntityPlayer player = ((NetHandlerPlayServer) event.handler).playerEntity; // EntityPlayerMP

        packetRx(stream, manager, player);
    }


    public void packetRx(DataInputStream stream, NetworkManager manager, EntityPlayer player) {
        try {
            switch (stream.readByte()) {
                case Eln.PACKET_PLAYER_KEY:
                    packetPlayerKey(stream, manager, player);
                    break;
                case Eln.PACKET_NODE_SINGLE_SERIALIZED:
                    packetNodeSingleSerialized(stream, manager, player);
                    break;
                case Eln.PACKET_PUBLISH_FOR_NODE:
                    packetForNode(stream, manager, player);
                    break;
                case Eln.PACKET_FOR_CLIENT_NODE:
                    packetForClientNode(stream, manager, player);
                    break;
                case Eln.PACKET_OPEN_LOCAL_GUI:
                    packetOpenLocalGui(stream, manager, player);
                    break;
                case Eln.PACKET_PLAY_SOUND:
                    packetPlaySound(stream, manager, player);
                    break;
                case Eln.PACKET_STOP_SOUND:

                    break;
                case Eln.PACKET_CLIENT_TO_SERVER_CONNECTION:
                    packetNewClient(manager, player);
                    break;
                case Eln.PACKET_SERVER_TO_CLIENT_INFO:
                    packetServerInfo(stream, manager, player);
                    break;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }


    private void packetNewClient(NetworkManager manager, EntityPlayer player) {

        ByteArrayOutputStream bos = new ByteArrayOutputStream(64);
        DataOutputStream stream = new DataOutputStream(bos);

        try {
            stream.writeByte(Eln.PACKET_SERVER_TO_CLIENT_INFO);
            for (IConfigSharing c : Eln.configShared) {
                c.serializeConfig(stream);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        Utils.sendPacketToClient(bos, (EntityPlayerMP) player);
    }

    private void packetServerInfo(DataInputStream stream, NetworkManager manager, EntityPlayer player) {
        for (IConfigSharing c : Eln.configShared) {
            try {
                c.deserialize(stream);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    void packetPlaySound(DataInputStream stream, NetworkManager manager, EntityPlayer player) {
        try {
            if (stream.readByte() != player.dimension)
                return;
            DP.println(DPType.SOUND, "Playing sound on client!");

            SoundCommand sc = SoundCommand.Companion.fromStream(stream, player.worldObj);
            ClientSoundHandler.SoundData sd = SoundCommand.Companion.toSoundData(sc, player);
            ClientProxy.clientSoundHandler.start(sd);

        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    void packetStopSound(DataInputStream stream, NetworkManager manager, EntityPlayer player) {

    }

    void packetOpenLocalGui(DataInputStream stream, NetworkManager manager, EntityPlayer player) {
        EntityPlayer clientPlayer = (EntityPlayer) player;
        try {
            clientPlayer.openGui(Eln.instance, stream.readInt(),
                clientPlayer.worldObj, stream.readInt(), stream.readInt(),
                stream.readInt());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    void packetForNode(DataInputStream stream, NetworkManager manager, EntityPlayer player) {
        try {
            Coordinate coordonate = new Coordinate(stream.readInt(),
                stream.readInt(), stream.readInt(), stream.readByte());

            NodeBase node = NodeManager.instance.getNodeFromCoordonate(coordonate);
            if (node != null && node.getNodeUuid().equals(stream.readUTF())) {
                node.networkUnserialize(stream, (EntityPlayerMP) player);
            } else {
                DP.println(DPType.NETWORK, "packetForNode node found");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    void packetForClientNode(DataInputStream stream, NetworkManager manager, EntityPlayer player) {
        EntityPlayer clientPlayer = (EntityPlayer) player;
        int x, y, z, dimention;
        try {

            x = stream.readInt();
            y = stream.readInt();
            z = stream.readInt();
            dimention = stream.readByte();


            if (clientPlayer.dimension == dimention) {
                TileEntity entity = clientPlayer.worldObj.getTileEntity(x, y, z);
                if (entity instanceof INodeEntity) {
                    INodeEntity node = (INodeEntity) entity;
                    if (node.getNodeUuid().equals(stream.readUTF())) {
                        node.serverPacketUnserialize(stream);
                        if (0 != stream.available()) {
                            DP.println(DPType.NETWORK, "0 != stream.available()");
                        }
                    } else {
                        DP.println(DPType.NETWORK, "Wrong node UUID warning");
                        int dataSkipLength = stream.readByte();
                        for (int idx = 0; idx < dataSkipLength; idx++) {
                            stream.readByte();
                        }
                    }
                }
            } else
                DP.println(DPType.NETWORK, "No node found for " + x + " " + y + " " + z);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    void packetNodeSingleSerialized(DataInputStream stream, NetworkManager manager, EntityPlayer player) {
        try {
            EntityPlayer clientPlayer = player;
            int x, y, z, dimention;
            x = stream.readInt();
            y = stream.readInt();
            z = stream.readInt();
            dimention = stream.readByte();

            if (clientPlayer.dimension == dimention) {
                TileEntity entity = clientPlayer.worldObj.getTileEntity(x, y, z);
                if (entity instanceof INodeEntity) {
                    INodeEntity node = (INodeEntity) entity;
                    if (node.getNodeUuid().equals(stream.readUTF())) {
                        node.serverPublishUnserialize(stream);
                        if (0 != stream.available()) {
                            DP.println(DPType.NETWORK, "0 != stream.available()");

                        }
                    } else {
                        DP.println(DPType.NETWORK, "Wrong node UUID warning");
                        int dataSkipLength = stream.readByte();
                        for (int idx = 0; idx < dataSkipLength; idx++) {
                            stream.readByte();
                        }
                    }
                } else
                    DP.println(DPType.NETWORK, "No node found for " + x + " " + y + " " + z);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * packetPlayerKey - handles key interactions
     * @param stream Network Stream
     * @param manager Network Manager
     * @param player Player reference (the one whose client sent the packet)
     */
    void packetPlayerKey(DataInputStream stream, NetworkManager manager, EntityPlayer player) {
        EntityPlayerMP playerMP = (EntityPlayerMP) player;
        try {
            byte id = stream.readByte();
            boolean state = stream.readBoolean();

            if (id == KeyRegistry.getKeyID("Wrench")) {
                PlayerManager.PlayerMetadata metadata = Eln.playerManager.get(playerMP);
                metadata.setInteractEnable(state);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/java/mods/eln/PacketHandler.kt
package mods.eln

import cpw.mods.fml.common.eventhandler.SubscribeEvent
import cpw.mods.fml.common.network.FMLNetworkEvent.ServerCustomPacketEvent
import io.netty.channel.ChannelHandler.Sharable
import mods.eln.client.ClientKeyHandler
import mods.eln.client.ClientProxy
import mods.eln.misc.Coordinate
import mods.eln.misc.Utils.println
import mods.eln.misc.Utils.sendPacketToClient
import mods.eln.node.INodeEntity
import mods.eln.node.NodeManager
import mods.eln.sound.SoundClient
import mods.eln.sound.SoundCommand
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.network.NetHandlerPlayServer
import net.minecraft.network.NetworkManager
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

@Sharable
class PacketHandler {
    @SubscribeEvent
    fun onServerPacket(event: ServerCustomPacketEvent) {
        val packet = event.packet
        val stream = DataInputStream(ByteArrayInputStream(packet.payload().array()))
        val manager = event.manager
        val player: EntityPlayer = (event.handler as NetHandlerPlayServer).playerEntity // EntityPlayerMP
        packetRx(stream, manager, player)
    }

    fun packetRx(stream: DataInputStream, manager: NetworkManager, player: EntityPlayer) {
        try {
            when (stream.readByte()) {
                Eln.packetPlayerKey -> packetPlayerKey(stream, manager, player)
                Eln.packetNodeSingleSerialized -> packetNodeSingleSerialized(stream, manager, player)
                Eln.packetPublishForNode -> packetForNode(stream, manager, player)
                Eln.packetForClientNode -> packetForClientNode(stream, manager, player)
                Eln.packetOpenLocalGui -> packetOpenLocalGui(stream, manager, player)
                Eln.packetPlaySound -> packetPlaySound(stream, manager, player)
                Eln.packetDestroyUuid -> packetDestroyUuid(stream, manager, player)
                Eln.packetClientToServerConnection -> packetNewClient(manager, player)
                Eln.packetServerToClientInfo -> packetServerInfo(stream, manager, player)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun packetNewClient(@Suppress("UNUSED_PARAMETER") manager: NetworkManager, player: EntityPlayer) {
        val bos = ByteArrayOutputStream(64)
        val stream = DataOutputStream(bos)
        try {
            stream.writeByte(Eln.packetServerToClientInfo.toInt())
            for (c in Eln.instance.configShared) {
                c.serializeConfig(stream)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        sendPacketToClient(bos, (player as EntityPlayerMP))
    }

    private fun packetServerInfo(stream: DataInputStream, @Suppress("UNUSED_PARAMETER") manager: NetworkManager, @Suppress("UNUSED_PARAMETER") player: EntityPlayer) {
        for (c in Eln.instance.configShared) {
            try {
                c.deserialize(stream)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    private fun packetDestroyUuid(stream: DataInputStream, @Suppress("UNUSED_PARAMETER") manager: NetworkManager, @Suppress("UNUSED_PARAMETER") player: EntityPlayer) {
        try {
            ClientProxy.uuidManager.kill(stream.readInt())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun packetPlaySound(stream: DataInputStream, @Suppress("UNUSED_PARAMETER") manager: NetworkManager, player: EntityPlayer) {
        try {
            if (stream.readByte().toInt() != player.dimension) return
            SoundClient.play(SoundCommand.fromStream(stream, player.worldObj))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun packetOpenLocalGui(stream: DataInputStream, @Suppress("UNUSED_PARAMETER") manager: NetworkManager, player: EntityPlayer) {
        try {
            player.openGui(Eln.instance, stream.readInt(),
                player.worldObj, stream.readInt(), stream.readInt(),
                stream.readInt())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun packetForNode(stream: DataInputStream, @Suppress("UNUSED_PARAMETER") manager: NetworkManager, player: EntityPlayer?) {
        try {
            val coordinate = Coordinate(stream.readInt(), stream.readInt(), stream.readInt(), stream.readByte().toInt())
            val node = NodeManager.instance!!.getNodeFromCoordonate(coordinate)
            if (node != null && node.nodeUuid == stream.readUTF()) {
                node.networkUnserialize(stream, player as EntityPlayerMP?)
            } else {
                println("packetForNode node found")
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun packetForClientNode(stream: DataInputStream, @Suppress("UNUSED_PARAMETER") manager: NetworkManager, player: EntityPlayer) {
        try {
            val x = stream.readInt()
            val y = stream.readInt()
            val z = stream.readInt()
            val dimension = stream.readByte().toInt()
            if (player.dimension == dimension) {
                val entity = player.worldObj.getTileEntity(x, y, z)
                if (entity != null && entity is INodeEntity) {
                    val node = entity as INodeEntity
                    if (node.nodeUuid == stream.readUTF()) {
                        node.serverPacketUnserialize(stream)
                        if (0 != stream.available()) {
                            println("0 != stream.available()")
                        }
                    } else {
                        println("Wrong node UUID warning")
                        val dataSkipLength = stream.readByte().toInt()
                        for (idx in 0 until dataSkipLength) {
                            stream.readByte()
                        }
                    }
                }
            } else println("No node found for $x $y $z")
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun packetNodeSingleSerialized(stream: DataInputStream, @Suppress("UNUSED_PARAMETER") manager: NetworkManager, player: EntityPlayer) {
        try {
            val x: Int = stream.readInt()
            val y: Int = stream.readInt()
            val z: Int = stream.readInt()
            val dimension: Int = stream.readByte().toInt()
            if (player.dimension == dimension) {
                val entity = player.worldObj.getTileEntity(x, y, z)
                if (entity != null && entity is INodeEntity) {
                    val node = entity as INodeEntity
                    if (node.nodeUuid == stream.readUTF()) {
                        node.serverPublishUnserialize(stream)
                        if (0 != stream.available()) {
                            println("0 != stream.available()")
                        }
                    } else {
                        println("Wrong node UUID warning")
                        val dataSkipLength = stream.readByte().toInt()
                        for (idx in 0 until dataSkipLength) {
                            stream.readByte()
                        }
                    }
                } else println("No node found for $x $y $z")
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun packetPlayerKey(stream: DataInputStream, @Suppress("UNUSED_PARAMETER") manager: NetworkManager, player: EntityPlayer?) {
        val playerMP = player as EntityPlayerMP?
        val id: Byte
        try {
            id = stream.readByte()
            val state = stream.readBoolean()
            if (id.toInt() == ClientKeyHandler.wrenchId) {
                val metadata = Eln.playerManager[playerMP!!]
                metadata!!.interactEnable = state
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    init {
        Eln.eventChannel.register(this)
    }
}
